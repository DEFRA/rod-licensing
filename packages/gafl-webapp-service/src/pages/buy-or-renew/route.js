import { BUY_OR_RENEW } from '../../uri.js'
import Joi from 'joi'
import pageRoute from '../../routes/page-route.js'
import { nextPage } from '../../routes/next-page.js'
export const buyOrRenew = {
  buy: 'buy',
  renew: 'renew'
}

export const validator = Joi.object({
  'buy-or-renew': Joi.string()
    .valid('buy-licence', 'renew-licence')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

const route = pageRoute(BUY_OR_RENEW.page, BUY_OR_RENEW.uri, validator, nextPage)

export default route
