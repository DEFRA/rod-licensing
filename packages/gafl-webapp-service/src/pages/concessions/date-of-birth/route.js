import { DATE_OF_BIRTH } from '../../../uri.js'
import Joi from 'joi'
import pageRoute from '../../../routes/page-route.js'
import { validation } from '@defra-fish/business-rules-lib'
import { nextPage } from '../../../routes/next-page.js'

export const validator = payload => {
  const dateOfBirth = `${payload['date-of-birth-year']}-${payload['date-of-birth-month']}-${payload['date-of-birth-day']}`
  Joi.assert(
    { 'date-of-birth': dateOfBirth },
    Joi.object({
      'date-of-birth': validation.contact.createBirthDateValidator(Joi)
    })
  )
}

export const getData = async request => {
  const { isLicenceForYou } = await request.cache().helpers.status.getCurrentPermission()

  return { isLicenceForYou }
}

export default pageRoute(DATE_OF_BIRTH.page, DATE_OF_BIRTH.uri, validator, nextPage, getData)
