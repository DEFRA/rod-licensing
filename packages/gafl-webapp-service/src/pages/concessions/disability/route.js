import { DISABILITY_CONCESSION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import * as concessionHelper from '../../../processors/concession-helper.js'
import { nextPage } from '../../../routes/next-page.js'
import { disabilityConcessionTypes } from './update-transaction.js'

const validator = Joi.object({
  'disability-concession': Joi.string()
    .valid(...Object.values(disabilityConcessionTypes))
    .required(),
  'ni-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.pipDla,
    then: validation.contact.createNationalInsuranceNumberValidator(Joi).required(),
    otherwise: Joi.string().empty('')
  }),
  'blue-badge-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.blueBadge,
    then: Joi.string()
      .max(25)
      .required(),
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    hasJunior: concessionHelper.hasJunior(permission),
    hasSenior: concessionHelper.hasSenior(permission),
    isLicenceForYou: permission.isLicenceForYou,
    ...disabilityConcessionTypes
  }
}

export default pageRoute(DISABILITY_CONCESSION.page, DISABILITY_CONCESSION.uri, validator, nextPage, getData)
