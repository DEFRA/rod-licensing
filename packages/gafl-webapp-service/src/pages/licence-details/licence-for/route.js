import Joi from 'joi'
import { LICENCE_FOR } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import { nextPage } from '../../../routes/next-page.js'

export const validator = Joi.object({
  'licence-for': Joi.string()
    .valid('you', 'someone-else')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

const route = pageRoute(LICENCE_FOR.page, LICENCE_FOR.uri, validator, nextPage)
const postRoute = route.find(r => r.method === 'POST')
const { handler } = postRoute
postRoute.handler = async (request, h) => {
  const { currentPermissionIdx } = await request.cache().helpers.status.get()
  const pageCache = Object.assign({}, await request.cache().helpers.page.get())
  const permission = await request.cache().helpers.page.getCurrentPermission(LICENCE_FOR.page)
  const licenceForHasChanged = (permission !== undefined) && permission.payload[LICENCE_FOR.page] !== request.payload[LICENCE_FOR.page]
  if (licenceForHasChanged) {
    pageCache.permissions[currentPermissionIdx] = {
      [LICENCE_FOR.page]: permission
    }
    await request.cache().helpers.page.set(pageCache)
  }
  return handler(request, h)
}
export default route
