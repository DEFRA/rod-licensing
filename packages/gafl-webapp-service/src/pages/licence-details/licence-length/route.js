import { LICENCE_LENGTH } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { pricingDetail } from '../../../processors/pricing-summary.js'
import { licenceTypeDisplay } from '../../../processors/licence-type-display.js'
import { nextPage } from '../../../routes/next-page.js'

const validator = Joi.object({
  'licence-length': Joi.string()
    .valid('12M', '8D', '1D')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const pricing = await pricingDetail(LICENCE_LENGTH.page, permission)
  console.log('Husky test')

  return {
    pricing,
    isLicenceForYou: permission.isLicenceForYou,
    licenceTypeStr: licenceTypeDisplay(permission)
  }
}

export default pageRoute(LICENCE_LENGTH.page, LICENCE_LENGTH.uri, validator, nextPage, getData)
