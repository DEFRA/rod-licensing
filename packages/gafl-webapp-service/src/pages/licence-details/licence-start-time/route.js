import { LICENCE_START_TIME } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import { SERVICE_LOCAL_TIME } from '@defra-fish/business-rules-lib'
import Joi from 'joi'
import moment from 'moment-timezone'
import { cacheDateFormat } from '../../../processors/date-and-time-display.js'
import { nextPage } from '../../../routes/next-page.js'

const getMinHour = permission => {
  const now = moment().tz(SERVICE_LOCAL_TIME)
  const permissionStartsToday = moment(permission.licenceStartDate, cacheDateFormat).tz(SERVICE_LOCAL_TIME).isSame(now, 'day')
  if (permissionStartsToday) {
    const cantStartUntilTomorrow = moment(now).add(90, 'minute').isAfter(now, 'day')
    if (cantStartUntilTomorrow) {
      return 24
    }
    return now.add(90, 'minute').startOf('hour').hour()
  }
  return 0
}

const hours = Array(25)
  .fill(0)
  .map((e, idx) => idx.toString())

const validator = (payload, options) => {
  const { permission } = options.context.app.request
  const minHour = getMinHour(permission)

  Joi.assert(
    payload,
    Joi.object({
      'licence-start-time': Joi.string()
        .valid(...hours.filter(h => h >= minHour))
        .required()
    }).options({ abortEarly: false, allowUnknown: true })
  )
}

const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const startDateStr = moment(permission.licenceStartDate, cacheDateFormat).format('dddd, Do MMMM, YYYY')

  return { startDateStr, minHour: getMinHour(permission) }
}

const route = pageRoute(LICENCE_START_TIME.page, LICENCE_START_TIME.uri, validator, nextPage, getData)
route.find(r => r.method === 'POST').options.ext = {
  onPostAuth: {
    method: async (request, reply) => {
      const permission = await request.cache().helpers.transaction.getCurrentPermission()
      request.app.permission = permission
      return reply.continue
    }
  }
}

export default route
