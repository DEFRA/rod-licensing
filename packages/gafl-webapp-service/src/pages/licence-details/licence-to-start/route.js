import Joi from 'joi'
import moment from 'moment-timezone'

import JoiDate from '@hapi/joi-date'
import { START_AFTER_PAYMENT_MINUTES, ADVANCED_PURCHASE_MAX_DAYS, SERVICE_LOCAL_TIME } from '@defra-fish/business-rules-lib'
import { LICENCE_TO_START } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import { dateFormats } from '../../../constants.js'
import { nextPage } from '../../../routes/next-page.js'

const JoiX = Joi.extend(JoiDate)

const validator = payload => {
  const licenceStartDate = `${payload['licence-start-date-year']}-${payload['licence-start-date-month']}-${payload['licence-start-date-day']}`
  Joi.assert(
    {
      'licence-start-date': licenceStartDate,
      'licence-to-start': payload['licence-to-start']
    },
    Joi.object({
      'licence-to-start': Joi.string()
        .valid('after-payment', 'another-date')
        .required(),
      'licence-start-date': Joi.alternatives().conditional('licence-to-start', {
        is: 'another-date',
        then: JoiX.date()
          .format(dateFormats)
          .min(
            moment()
              .tz(SERVICE_LOCAL_TIME)
              .startOf('day')
          )
          .max(
            moment()
              .tz(SERVICE_LOCAL_TIME)
              .add(ADVANCED_PURCHASE_MAX_DAYS, 'days')
          )
          .required(),
        otherwise: Joi.string().empty('')
      })
    }).options({ abortEarly: false, allowUnknown: true })
  )
}

export const getData = async request => {
  const fmt = 'DD MM YYYY'
  const { isLicenceForYou } = await request.cache().helpers.status.getCurrentPermission()

  return {
    isLicenceForYou,
    exampleStartDate: moment()
      .tz(SERVICE_LOCAL_TIME)
      .add(1, 'days')
      .format(fmt),
    minStartDate: moment()
      .tz(SERVICE_LOCAL_TIME)
      .format(fmt),
    maxStartDate: moment()
      .tz(SERVICE_LOCAL_TIME)
      .add(ADVANCED_PURCHASE_MAX_DAYS, 'days')
      .format(fmt),
    advancedPurchaseMaxDays: ADVANCED_PURCHASE_MAX_DAYS,
    startAfterPaymentMinutes: START_AFTER_PAYMENT_MINUTES
  }
}

export default pageRoute(LICENCE_TO_START.page, LICENCE_TO_START.uri, validator, nextPage, getData)
