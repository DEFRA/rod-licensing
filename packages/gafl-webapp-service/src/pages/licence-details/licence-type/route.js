import { LICENCE_TYPE, FRESHWATER_FISING_RULES, LOCAL_BYELAWS } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import { pricingDetail } from '../../../processors/pricing-summary.js'
import Joi from 'joi'
import * as concessionHelper from '../../../processors/concession-helper.js'
import { nextPage } from '../../../routes/next-page.js'

export const licenseTypes = {
  troutAndCoarse2Rod: 'trout-and-coarse-2-rod',
  troutAndCoarse3Rod: 'trout-and-coarse-3-rod',
  salmonAndSeaTrout: 'salmon-and-sea-trout'
}

export const validator = Joi.object({
  'licence-type': Joi.string()
    .valid(...Object.values(licenseTypes))
    .required()
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const pricing = await pricingDetail(LICENCE_TYPE.page, permission)

  return {
    licenseTypes,
    permission,
    pricing,
    isLicenceForYou: permission.isLicenceForYou,
    hasJunior: concessionHelper.hasJunior(permission),
    uri: {
      freshWaterFishingRules: FRESHWATER_FISING_RULES.uri,
      localByelaws: LOCAL_BYELAWS.uri,
      catchReturn: CATCH_RETURN.uri
    }
  }
}

export default pageRoute(LICENCE_TYPE.page, LICENCE_TYPE.uri, validator, nextPage, getData)
