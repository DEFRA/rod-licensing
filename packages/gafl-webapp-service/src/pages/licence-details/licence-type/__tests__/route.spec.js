import { getData, validator } from '../route'
import pageRoute from '../../../../routes/page-route.js'
import { nextPage } from '../../../../routes/next-page.js'
import '../../../../processors/pricing-summary.js'

jest.mock('../../../../routes/next-page.js')
jest.mock('../../../../routes/page-route.js')
jest.mock('../../../../processors/pricing-summary.js')

describe('licence-type > route', () => {
  const mockTransactionCacheGet = jest.fn()

  const mockRequest = {
    cache: () => ({
      helpers: {
        transaction: {
          getCurrentPermission: mockTransactionCacheGet
        }
      }
    })
  }

  describe('getData', () => {
    it('should return isLicenceForYou as true, if isLicenceForYou is true on the transaction cache', async () => {
      mockTransactionCacheGet.mockImplementationOnce(() => ({ concessions: [], isLicenceForYou: true }))
      const result = await getData(mockRequest)
      expect(result.isLicenceForYou).toBeTruthy()
    })

    it('should return isLicenceForYou as false, if isLicenceForYou is false on the transaction cache', async () => {
      mockTransactionCacheGet.mockImplementationOnce(() => ({ concessions: [], isLicenceForYou: false }))
      const result = await getData(mockRequest)
      expect(result.isLicenceForYou).toBeFalsy()
    })
  })

  describe('default', () => {
    it('should call the pageRoute with licence-type, /buy/licence-type, validator and nextPage', async () => {
      expect(pageRoute).toBeCalledWith('licence-type', '/buy/licence-type', validator, nextPage, getData)
    })
  })
})
