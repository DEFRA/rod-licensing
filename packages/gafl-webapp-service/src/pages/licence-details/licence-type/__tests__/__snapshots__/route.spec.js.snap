// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`licence-type > route getData getData returns expected output 1`] = `
Object {
  "hasJunior": true,
  "isLicenceForYou": undefined,
  "licenseTypes": Object {
    "salmonAndSeaTrout": "salmon-and-sea-trout",
    "troutAndCoarse2Rod": "trout-and-coarse-2-rod",
    "troutAndCoarse3Rod": "trout-and-coarse-3-rod",
  },
  "permission": Symbol(mock permission),
  "pricing": "test",
  "uri": Object {
    "freshWaterFishingRules": "https://www.gov.uk/freshwater-rod-fishing-rules",
    "localByelaws": "https://www.gov.uk/government/collections/local-fishing-byelaws",
  },
}
`;
